eventmesh-connectors/eventmesh-connector-rocketmq/build.gradle

```diff
@@ -16,21 +16,25 @@
  */
 
 List rocketmq = [
-        "org.apache.rocketmq:rocketmq-client:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-broker:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-common:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-store:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-namesrv:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-tools:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-remoting:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-logging:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-srvutil:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-filter:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-acl:$rocketmq_version",
-        "org.apache.rocketmq:rocketmq-srvutil:$rocketmq_version",
-
+        "org.apache.rocketmq:rocketmq-client:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-broker:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-common:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-store:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-namesrv:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-tools:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-remoting:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-logging:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-srvutil:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-filter:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-acl:${rocketmq_version}",
+        "org.apache.rocketmq:rocketmq-srvutil:${rocketmq_version}"
 ]
 
+configurations {
+    implementation.exclude group: "ch.qos.logback", module: "logback-classic"
+    implementation.exclude group: "ch.qos.logback", module: "logback-core"
+}
```

Comment:
I'm a little confused as to why some modules have removed the exclusion statements for the Logging backend and others have added them.

Is it better to exclude them uniformly in the root directory, or to exclude them where the dependencies are introduced?

Reply:
There are two ways logging backends appear in EventMesh artifacts:

- as **direct** dependencies in your modules. I have removed all of those, the `configurations { runtimeClasspath {} }` block in the root project should just prevent reintroducing them by mistake,
- as **transitive** dependencies of your **direct** dependencies. For example [`rocketmq-broker`](https://mvnrepository.com/artifact/org.apache.rocketmq/rocketmq-broker/4.9.5) seems to be such a dependency. Since it depends on Logback, all your artifacts that use RocketMQ effectively depend on Logback.

I tested the second point by running:
```sh
./gradlew generatePomFileForMavenJava
```
and compared the generated `eventmesh-storage-plugin/eventmesh-storage-rocketmq/build/publications/mavenJava/pom-default.xml` files. The POM file **before** the introduction of these exclusions contained just:
```xml
  <dependencies>
    ...
    <dependency>
      <groupId>org.apache.rocketmq</groupId>
      <artifactId>rocketmq-broker</artifactId>
      <version>4.9.5</version>
      <scope>runtime</scope>
    </dependency>
    ...
  </dependencies>
```

After the changes it contained exclusions:
```xml
  <dependencies>
    ...
    <dependency>
      <groupId>org.apache.rocketmq</groupId>
      <artifactId>rocketmq-broker</artifactId>
      <version>4.9.5</version>
      <scope>runtime</scope>
      <exclusions>
        <exclusion>
          <artifactId>logback-classic</artifactId>
          <groupId>ch.qos.logback</groupId>
        </exclusion>
        <exclusion>
          <artifactId>logback-core</artifactId>
          <groupId>ch.qos.logback</groupId>
        </exclusion>
      </exclusions>
    </dependency>
    ...
  </dependencies>
```

Reply:
If `configurations { runtimeClasspath {} }` block in the root project should be able to avoid the transitive logging backend passed by direct dependencies globally, then there's no need to exclude Logback again where the RocketMQ dependencies are introduced.

Some `implementation.exclude` statements are removed (like `eventmesh-connectors/eventmesh-connector-rabbitmq/build.gradle`) and I think that's because `com.rabbitmq:amqp-client` won't introduce any logging backend and these statements are redundant, right?

Reply:
I did some experiments and the `configuration { runtimeClasspath {} }` block might prevent the propagation of logging backends in the EventMesh project, but it is ineffective for third-party projects.

A better alternative would be the [Gradle Enforcer Plugin](https://kordamp.org/enforcer-gradle-plugin/) configured with a [`BannedDependencies`](https://kordamp.org/enforcer-gradle-plugin/#_banneddependencies) rule. Unfortunately I was not able to configure it without triggering the kordamp/enforcer-gradle-plugin#22 issue.

In the end, I removed the `configurations` block entirely and replaced it with surgical strikes against broken dependencies (RocketMQ, Pulsar and Zookeeper). I also excluded `spring-boot-starter-logging` from Spring Boot.

Reply:
I love your use of "surgical strikes"😄